File: packages/shared/src/events/events.test.ts
import { describe, expect, it } from 'vitest';

import { IPC_CHANNELS } from '../ipc/contracts/index';
import {
  type AppEvents,
  EVENT_CHANNELS,
  isPipelineSectionKey,
  PIPELINE_STAGE_PHASES,
  type PipelineSectionKey,
  type PipelineStageEvent,
  SECTION_KEY_MAX_LENGTH,
} from './index';

/** Flatten every event-channel wire string out of the namespaced registry. */
function flattenEventChannels(): string[] {
  const all: string[] = [];
  for (const namespace of Object.values(EVENT_CHANNELS)) {
    all.push(...(Object.values(namespace) as string[]));
  }
  return all;
}

/** Flatten every request/response wire string out of `IPC_CHANNELS`. */
function flattenIpcChannels(): string[] {
  const all: string[] = [];
  for (const namespace of Object.values(IPC_CHANNELS)) {
    all.push(...(Object.values(namespace) as string[]));
  }
  return all;
}

describe('EVENT_CHANNELS', () => {
  it('exposes every expected event namespace', () => {
    const namespaces = Object.keys(EVENT_CHANNELS).sort();
    expect(namespaces).toEqual([
      'ai',
      'applications',
      'autopilot',
      'boards',
      'extensionBridge',
      'jobs',
      'menu',
      'notifications',
      'pipeline',
      'scrape',
      'system',
      'updater',
    ]);
  });

  it('maps every event channel to a non-empty string', () => {
    for (const namespace of Object.values(EVENT_CHANNELS)) {
      for (const channel of Object.values(namespace)) {
        expect(typeof channel).toBe('string');
        expect((channel as string).length).toBeGreaterThan(0);
      }
    }
  });

  it('has globally unique event-channel identifiers', () => {
    const all = flattenEventChannels();
    expect(new Set(all).size).toBe(all.length);
  });

  it('never collides with any IPC request/response channel', () => {
    // Events and IPC channels share the same Tauri wire namespace, so a name
    // reused as both a command and a push event would be ambiguous. Assert the
    // two flattened sets are disjoint.
    const eventChannels = new Set(flattenEventChannels());
    const ipcChannels = flattenIpcChannels();
    const intersection = ipcChannels.filter((c) => eventChannels.has(c));
    expect(intersection).toEqual([]);
  });

  it('keeps AppEvents and EVENT_CHANNELS in 1:1 sync', () => {
    // `AppEvents` is keyed by wire name -> payload type. Every registry value
    // must have a payload entry and vice-versa, so the registry and the typed
    // payload map can never drift.
    const appEventKeys: Array<keyof AppEvents> = [
      'ai:stream',
      'jobs:event',
      'applications:changed',
      'notifications:changed',
      'notifications:open',
      'notifications:toast',
      'updater:status',
      'menu:navigate',
      'menu:action',
      'autopilot:focus',
      'autopilot:step',
      'pipeline:stage',
      'scrape:progress',
      'scrape:item',
      'boards:login-status',
      // Windows accent-change push (WinRT UISettings::ColorValuesChanged).
      'system:accentChanged',
      'extensionBridge:changed',
    ];
    const eventChannels = flattenEventChannels().sort();
    expect((appEventKeys as string[]).slice().sort()).toEqual(eventChannels);
  });
});

describe('PipelineStageEvent', () => {
  /**
   * Wire-shape lock for `pipeline:stage`. The Rust emitter (Phase 3) and this
   * interface are hand-synced, so the field NAMES are the contract: renaming
   * one silently breaks the renderer with no type error on the Rust side.
   * The `satisfies` annotation catches a required field being added or removed;
   * the key assertion catches a rename or a casing slip.
   */
  it('pins the full camelCase field set', () => {
    const full = {
      runId: 'run-1',
      jobId: 'job-1',
      stage: 'draft',
      phase: 'finish',
      index: 2,
      total: 6,
      attempt: 1,
      sectionKey: 'experience:0',
      ms: 1234,
      issueCount: 3,
      criticalCount: 1,
    } satisfies PipelineStageEvent;

    expect(Object.keys(full).sort()).toEqual(
      [
        'attempt',
        'criticalCount',
        'index',
        'issueCount',
        'jobId',
        'ms',
        'phase',
        'runId',
        'sectionKey',
        'stage',
        'total',
      ].sort()
    );
  });

  /**
   * The five optional fields really are optional: a `start` event has no
   * duration, no issue counts, and (for a non-section stage) no section key.
   * A required-by-accident field would strand the emitter with nothing to send.
   */
  it('accepts a start event with only the required fields', () => {
    const start = {
      runId: 'run-1',
      jobId: 'job-1',
      stage: 'plan',
      phase: 'start',
      index: 0,
      total: 6,
      attempt: 1,
    } satisfies PipelineStageEvent;

    expect(Object.keys(start).sort()).toEqual([
      'attempt',
      'index',
      'jobId',
      'phase',
      'runId',
      'stage',
      'total',
    ]);
  });

  /**
   * The phase vocabulary is closed — three values, no free-form strings.
   * Asserts against the EXPORTED array (the type's own source of truth), not
   * a literal duplicated here, so a phase added to the array without a
   * corresponding update to this expectation fails — the type itself cannot
   * drift from the array since it is derived from it.
   */
  it('pins the phase vocabulary', () => {
    const phases: readonly PipelineStageEvent['phase'][] = PIPELINE_STAGE_PHASES;
    expect(phases).toEqual(['start', 'finish', 'error']);
  });

  /**
   * `sectionKey` is the ONE model-derived field on a channel that claims to be
   * content-free, so its vocabulary is closed exactly like `phase`'s and the
   * guard is the checkable form of that claim. These cases are the contract the
   * Phase-3 Rust emitter must satisfy before anything reaches the wire.
   */
  describe('sectionKey', () => {
    it('accepts every form of the closed grammar', () => {
      for (const key of [
        'summary',
        'skills',
        'projects',
        'education',
        'experience:0',
        'experience:7',
        'experience:255',
      ]) {
        expect(isPipelineSectionKey(key), key).toBe(true);
      }
    });

    it('rejects a free-form label the model invented', () => {
      for (const key of [
        'Work History',
        'experience', // the index is not optional
        'experience:', // ...nor empty
        'experience:256', // ...nor wider than a u8
        'experience:01', // ...nor zero-padded
        'experience:1.5',
        'experience:-1',
        'experience:0; DROP TABLE runs',
        'SUMMARY', // the grammar is lower-case
        '',
      ]) {
        expect(isPipelineSectionKey(key), key).toBe(false);
      }
    });

    it('rejects an over-length value and anything that is not a string', () => {
      expect(isPipelineSectionKey(`experience:${'9'.repeat(10_000)}`)).toBe(false);
      expect(isPipelineSectionKey('summary'.padEnd(SECTION_KEY_MAX_LENGTH + 1, '!'))).toBe(false);
      for (const notAKey of [undefined, null, 3, {}, ['summary']]) {
        expect(isPipelineSectionKey(notAKey)).toBe(false);
      }
    });

    /** The bound must admit the longest LEGAL key, or the guard contradicts the
     *  grammar it is enforcing. */
    it('bounds the wire without excluding a legal key', () => {
      expect('experience:255'.length).toBeLessThanOrEqual(SECTION_KEY_MAX_LENGTH);
    });

    /**
     * The vocabulary is closed at the TYPE level as well, so a Phase-3 renderer
     * consumer gets the same guarantee statically. `@ts-expect-error` is checked
     * by `tsc` (not by vitest's esbuild), and fails if the assignment ever
     * becomes legal — i.e. if `sectionKey` is widened back to `string`.
     */
    it('closes the vocabulary at the type level, not just at runtime', () => {
      // @ts-expect-error — a free-form label the model invented is not a key.
      const invented: PipelineSectionKey = 'Work History';
      expect(isPipelineSectionKey(invented)).toBe(false);
    });
  });
});

File: apps/desktop/src-tauri/src/ipc_contracts/events_test.rs
//! Pins for the generated `pipeline:stage` vocabulary consts.
//!
//! `events.rs` is @generated/DO-NOT-EDIT, so these are hand-written. `pnpm
//! gen:ipc:check` proves the file MATCHES the TS source of truth; it cannot
//! prove the source of truth is self-consistent. That is what these check —
//! on the RUST consts, because they are what the Phase-3 emitter validates
//! against (the TS twins are covered by `packages/shared/src/events/events.test.ts`).

use super::events::{
    is_pipeline_section_key, PIPELINE_SECTION_EXPERIENCE_PREFIX, PIPELINE_SECTION_KEYS_FIXED,
    PIPELINE_STAGE_PHASES, SECTION_KEY_MAX_LENGTH,
};

/// The longest key the grammar can produce is `experience:255`. If the bound
/// ever drops below it, the emitter would reject a value its own grammar calls
/// legal — the guard contradicting the contract it enforces.
#[test]
fn the_bound_admits_the_longest_legal_section_key() {
    let longest = format!("{PIPELINE_SECTION_EXPERIENCE_PREFIX}255");
    assert!(
        longest.chars().count() <= SECTION_KEY_MAX_LENGTH,
        "'{longest}' is the longest legal key but exceeds SECTION_KEY_MAX_LENGTH \
         ({SECTION_KEY_MAX_LENGTH})"
    );
    for key in PIPELINE_SECTION_KEYS_FIXED {
        assert!(
            key.chars().count() <= SECTION_KEY_MAX_LENGTH,
            "fixed key '{key}' exceeds SECTION_KEY_MAX_LENGTH ({SECTION_KEY_MAX_LENGTH})"
        );
    }
}

/// Value pin: the consts must carry the grammar, not some other vocabulary that
/// happens to be the right shape (both halves are short lower-case strings, so a
/// codegen mix-up with `PIPELINE_STAGE_PHASES` would satisfy every length check
/// above). Widening the grammar is a deliberate contract change — update the TS
/// source of truth, regenerate, and move this pin with it.
#[test]
fn the_section_key_grammar_is_pinned_to_its_literals() {
    assert_eq!(
        PIPELINE_SECTION_KEYS_FIXED,
        &["summary", "skills", "projects", "education"]
    );
    assert_eq!(PIPELINE_SECTION_EXPERIENCE_PREFIX, "experience:");
    assert_eq!(SECTION_KEY_MAX_LENGTH, 24);
}

/// The two halves must not overlap: a fixed key that also parses as the indexed
/// form (or an empty prefix, which would make EVERY key indexed) turns a closed
/// grammar into an ambiguous one.
#[test]
fn the_two_halves_of_the_grammar_are_disjoint_and_non_empty() {
    assert!(!PIPELINE_SECTION_EXPERIENCE_PREFIX.is_empty());
    assert!(!PIPELINE_SECTION_KEYS_FIXED.is_empty());
    assert!(!PIPELINE_STAGE_PHASES.is_empty());
    for key in PIPELINE_SECTION_KEYS_FIXED {
        assert!(
            !key.starts_with(PIPELINE_SECTION_EXPERIENCE_PREFIX),
            "fixed key '{key}' also matches the indexed form"
        );
    }
}

/// The generated guard must accept exactly what the TS `isPipelineSectionKey`
/// accepts — the cases mirror `packages/shared/src/events/events.test.ts`.
#[test]
fn the_guard_admits_the_grammar_and_nothing_else() {
    for key in [
        "summary",
        "skills",
        "projects",
        "education",
        "experience:0",
        "experience:7",
        "experience:255",
    ] {
        assert!(
            is_pipeline_section_key(key),
            "'{key}' is legal grammar but the guard rejected it"
        );
    }
    for key in [
        "Work History",
        "experience",     // the index is not optional
        "experience:",    // ...nor empty
        "experience:256", // ...nor wider than a u8
        "experience:1.5",
        "experience:-1",
        "experience:0; DROP TABLE runs",
        "SUMMARY", // the grammar is lower-case
        "",
    ] {
        assert!(
            !is_pipeline_section_key(key),
            "'{key}' is not the grammar but the guard accepted it"
        );
    }
    // Length is checked FIRST, so a value far past the bound is rejected without
    // the guard doing any further work on it.
    let hostile = format!("{PIPELINE_SECTION_EXPERIENCE_PREFIX}{}", "9".repeat(10_000));
    assert!(!is_pipeline_section_key(&hostile));
}

/// THE canonical-decimal pin. `str::parse::<u8>` is LOOSER than the TS regex
/// `^(0|[1-9][0-9]{0,2})$`: it accepts a `+` sign and leading zeros, so an
/// emitter that only parsed would put `experience:01` — a second spelling of
/// `experience:1` — on a wire whose vocabulary is supposed to be closed. Each
/// case asserts BOTH halves: that a bare parse would have let it through, and
/// that the guard does not. Loosening the generated guard back to a plain parse
/// fails here.
#[test]
fn the_indexed_half_admits_only_canonical_decimals() {
    for index in ["01", "007", "+1", "0255"] {
        assert!(
            index.parse::<u8>().is_ok(),
            "'{index}' no longer parses as a u8 — this pin has lost its point"
        );
        let key = format!("{PIPELINE_SECTION_EXPERIENCE_PREFIX}{index}");
        assert!(
            !is_pipeline_section_key(&key),
            "'{key}' is not canonical decimal but the guard accepted it — the \
             guard is parsing before validating"
        );
    }
    // Whitespace and the other non-canonical spellings a parse also rejects,
    // pinned so the guard can never start trimming its input.
    for index in [" 1", "1 ", "1_0", "٣"] {
        let key = format!("{PIPELINE_SECTION_EXPERIENCE_PREFIX}{index}");
        assert!(
            !is_pipeline_section_key(&key),
            "'{key}' is not canonical decimal but the guard accepted it"
        );
    }
}
